hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/upgrade/DataNodeUpgradeFinalizer.java

```diff
@@ -0,0 +1,99 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.container.upgrade;
+
+import static org.apache.hadoop.ozone.upgrade.UpgradeFinalizer.Status.FINALIZATION_DONE;
+import static org.apache.hadoop.ozone.upgrade.UpgradeFinalizer.Status.FINALIZATION_IN_PROGRESS;
+import static org.apache.hadoop.ozone.upgrade.UpgradeFinalizer.Status.FINALIZATION_REQUIRED;
+
+import java.io.IOException;
+import java.util.concurrent.Callable;
+
+import org.apache.hadoop.hdds.upgrade.HDDSLayoutFeatureCatalog.HDDSLayoutFeature;
+import org.apache.hadoop.ozone.container.common.statemachine.DatanodeStateMachine;
+import org.apache.hadoop.ozone.upgrade.BasicUpgradeFinalizer;
+
+/**
+ * UpgradeFinalizer for the DataNode.
+ */
+public class DataNodeUpgradeFinalizer extends
+    BasicUpgradeFinalizer<DatanodeStateMachine, DataNodeLayoutVersionManager> {
+
+  public DataNodeUpgradeFinalizer(DataNodeLayoutVersionManager versionManager) {
+    super(versionManager);
+  }
+
+  @Override
+  public StatusAndMessages finalize(String upgradeClientID,
+                                    DatanodeStateMachine dsm)
+      throws IOException {
+    StatusAndMessages response = preFinalize(upgradeClientID, dsm);
+    if (response.status() != FINALIZATION_REQUIRED) {
+      return response;
+    }
+    new Worker(dsm).call();
+    return STARTING_MSG;
+  }
+
+  private class Worker implements Callable<Void> {
+    private DatanodeStateMachine datanodeStateMachine;
+
+    /**
+     * Initiates the Worker, for the specified DataNode instance.
+     * @param dsm the DataNodeStateMachine instance on which to finalize the
+     *           new LayoutFeatures.
+     */
+    Worker(DatanodeStateMachine dsm) {
+      datanodeStateMachine = dsm;
+    }
+
+    @Override
+    public Void call() throws IOException {
+      if(!datanodeStateMachine.preFinalizeUpgrade()) {
+      // datanode is not yet ready to finalize.
+      // Reset the Finalization state.
+        versionManager.setUpgradeState(FINALIZATION_REQUIRED);
+        return null;
+      }
+      try {
+        emitStartingMsg();
+        versionManager.setUpgradeState(FINALIZATION_IN_PROGRESS);
+        /*
+         * Before we can call finalize the feature, we need to make sure that
+         * all existing pipelines are closed and pipeline Manger would freeze
+         * all new pipeline creation.
+         */
+
+        for (HDDSLayoutFeature f : versionManager.unfinalizedFeatures()) {
+          finalizeFeature(f, datanodeStateMachine.getDataNodeStorageConfig());
```

Comment:
We currently have a common layout feature hierarchy for SCM and Datanode. It is likely that a feature can either be part of the SCM or the DN. If we call on the onFinalize action in both the SCM and DN for every unfinalized HDDS layout feature, then we may be left with unintended consequences of trying to call on finalize action on an SCM feature on the DN. We should either have a marker on every layout feature to say if it belongs to SCM or DN, or not support finalization actions for DN for the V1 upgrade implementation.

Reply:
yes. I added separate SCM and datanode actions. Please take a look.